### PR TITLE
feat(new_split_chunks): support `maxInitialRequests` and `maxAsyncRequests` internally

### PR DESCRIPTION
**Pull request by hyf0**
## Summary

- I will add tests on the PR that exposes `maxInitialRequests` and `maxAsyncRequests` option

### <samp>🤖 Generated by Copilot at 4624d05</samp>

This pull request adds support for limiting the number of requests per chunk and respecting the minimum number of chunks per module group for the `splitChunks` option of the `SplitChunksPlugin`. It modifies the `plugin`, `cache_group`, and `options` modules of the `rspack_plugin_split_chunks_new` and `rspack_binding_options` crates, and adds a new `max_request` module. It also updates the changelog and version for the `@rspack/binding` package.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 4624d05</samp>

*  Add a markdown file to indicate a patch-level update for the `@rspack/binding` package ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-0b558caca3e3b7a355614a0f301954de345d7ca10911448e4af80adefe1ac0bbR1-R5))
*  Add `max_async_requests` and `max_initial_requests` fields to the `RawSplitChunks` struct in `raw_split_chunks.rs` and mark them as TODO for user exposure ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968R194-R197))
*  Add `max_initial_requests` and `max_async_requests` fields to the `CacheGroup` struct in `cache_group.rs` and store the values from the raw configuration or the defaults ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-0aab5b8bd0e625e143e1d9cfe68dbf4e7339d9646deffa103a03eaef98c527ebR32-R33))
*  Implement the `ensure_max_request_fit` method in the `max_request.rs` module to remove chunks that exceed the request limits from the cache group ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-df5e22b8430b27c0d6936f0c0a74a3f449f5b84a3b60d7c84eacb5fb719e88b9R1-R67))
*  Import the `max_request.rs` module into the `plugin` module in `plugin/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aR2))
*  Call the `ensure_max_request_fit` method and check the `min_chunks` requirement in the `apply` method of the `SplitChunksPlugin` struct in `plugin/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL65-R77))
*  Rename the variable `keys_of_empty_group` to `keys_of_invalid_group` and retain only the used chunks and valid module groups in the `remove_modules_from_other_entries_and_update_size` method of the `SplitChunksPlugin` struct in `plugin/module_group.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-491710d760a9177c165f505436fcb4f19f6f00caaeb2cd888f6377d881827d23L150-R150), [link](https://github.com/web-infra-dev/rspack/pull/3119/files?diff=unified&w=0#diff-491710d760a9177c165f505436fcb4f19f6f00caaeb2cd888f6377d881827d23L173-R197))

</details>

### CODE REVIEW

File: crates/rspack_plugin_split_chunks_new/src/plugin/max_request.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+        if actually_requests >= allowed_max_request {
+          Some(chunk.ukey)
+        } else {
+          None
+        }
+      })
+      .collect::<Vec<_>>();
+    invalided_chunks.into_iter().for_each(|c| {
+      used_chunks.to_mut().remove(&c);
```
Comment (h-a-n-a) [inline crates/rspack_plugin_split_chunks_new/src/plugin/max_request.rs:64]: Haven't dig into how webpack implements this feature. Could you describe more about what the relationship between `used_chunks` and `max_xxx_requests` is and how it is affected?

Comment (hyf0) [reply crates/rspack_plugin_split_chunks_new/src/plugin/max_request.rs:64]: I add comments at https://github.com/web-infra-dev/rspack/pull/3119/files/1fa4aa8b4bc5d87ef1121f575b8b92d7d1a843fb#diff-df5e22b8430b27c0d6936f0c0a74a3f449f5b84a3b60d7c84eacb5fb719e88b9R34-R45.

`Chunk`s in `used_chunks` are all code-splitting chunk, which means there are created by entry config or dynamic import.

If a code-splitting chunk is not split by `SplitChunksPlugin`, the number of requests for the chunk is 1.

If the code-splitting chunks is split by `SplitChunksPlugin`, to load the code-splitting chunk, with correct semantics, we need to also load the chunks derive from the the code-splitting chunk. Chunks derive from the the code-splitting chunk is in the same ChunkGroup with the split chunk.

So the number of requests is the length of `ChunkGroup#chunks` which belong to the split code-splitting chunk.

`max_xxx_requests` filter out chunks in `used_chunks` that doesn't satisfy `max_xxx_requests`.

You could see the described behavior at https://webpack.js.org/plugins/split-chunks-plugin/#splitchunksmaxinitialrequests.

Comment (hyf0) [reply crates/rspack_plugin_split_chunks_new/src/plugin/max_request.rs:64]: In general speaking, `max_xxx_requests` controls a code-splitting chunk(create in code splitting) that could be split how many times.

Comment (h-a-n-a) [reply crates/rspack_plugin_split_chunks_new/src/plugin/max_request.rs:64]: I fully understand the semantics of loading a module, however, what I don't understand is how removing `used_chunks` keeps the same semantics and when will the chunks removed from `used_chunks` be loaded.

**Comment by hyf0** [comment]
Some tests failed, I'm going to convert this PR to a draft.

**Review by h-a-n-a — APPROVED** [review]

**Pull request merged**